spyder/api/plugins.py

```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
-                          add_shortcut_to_tip=False):
-        """
-        Register a shortcut associated to a QAction or a QShortcut to
-        Spyder main application.
+    # Use this signal to display a message in the status bar.
+    # str: The message you want to display
+    # int: Amount of time to display the message
+    sig_show_status_message = Signal(str, int)
 
-        Parameters
-        ----------
-        qaction_or_qshortcut: QAction or QShortcut
-            QAction to register the shortcut for or QShortcut.
-        context: str
-            Name of the plugin this shortcut applies to. For instance,
-            if you pass 'Editor' as context, the shortcut will only
-            work when the editor is focused.
-            Note: You can use '_' if you want the shortcut to be work
-            for the entire application.
-        name: str
-            Name of the action the shortcut refers to (e.g. 'Debug
-            exit').
-        add_shortcut_to_tip: bool
-            If True, the shortcut is added to the action's tooltip.
-            This is useful if the action is added to a toolbar and
-            users hover it to see what it does.
-        """
-        super(BasePluginWidget, self)._register_shortcut(
-            qaction_or_qshortcut,
-            context,
-            name,
-            add_shortcut_to_tip)
+    # Use this signal to inform another plugin that configuration changed.
+    sig_option_changed = Signal(str, object)
 
-    def register_widget_shortcuts(self, widget):
-        """
-        Register shortcuts defined by a plugin's widget so they take
-        effect when the plugin is focused.
+    # TODO: DESCRIPTION?
+    sig_update_plugin_title = Signal()
 
-        Parameters
-        ----------
-        widget: QWidget
-            Widget to register shortcuts for.
+    # --- Private methods ----------------------------------------------------
+    # ------------------------------------------------------------------------
+    def __init__(self, parent, configuration=None):
+        super().__init__(parent)
+        self._initialize(parent, configuration)
 
-        Notes
-        -----
-        The widget interface must have a method called
-        `get_shortcut_data` for this to work. Please see
-        `spyder/widgets/findreplace.py` for an example.
-        """
-        for qshortcut, context, name in widget.get_shortcut_data():
-            self.register_shortcut(qshortcut, context, name)
+        # Dockwidget for the plugin, i.e. the pane that's going to be
+        # displayed in Spyder for this plugin.
+        # Note: This is created when you call the `add_dockwidget`
+        # method, which must be done in the `register_plugin` one.
+        #
+        # Names we use here: register, initialize, on_first_registration
+        # _initialize_plugin_in_mainwindow_layout
+        # Name we use in main window: setup, post_visible_setup
```

Comment:
* `_initialize` and `_initialize_plugin_in_mainwindow_layout` are hidden, so I see no problem how we name them.
* `register_plugin` and `on_first_registration` have been with us since Pierre's days. We could change their names, but as I said above, that will break a lot of third-party stuff not developed by us for little gain.

Reply:
>_initialize and _initialize_plugin_in_mainwindow_layout are hidden, so I see no problem how we name them.

Ok

>register_plugin and on_first_registration have been with us since Pierre's days. We could change their names

Doing things the way they always have been done for the sake of it is a really poor justification.

>but as I said above, that will break a lot of third-party stuff not developed by us for little gain

I do not plan to break compatibility between 4 and 5. (as was the case with 2-3, 3-4...)